### ISSUE TITLE
Nat type index misinterpreted

### ISSUE THREAD

**Issue by gergoerdi** [ucsd-progsys/liquidhaskell#2080]
```
{-# LANGUAGE DataKinds #-}
import Prelude
import GHC.TypeLits

data NatIndexed (n :: Nat)

{-@ type NI4 = NatIndexed 4 @-}

{-@ foo :: NI4 -> NI4 @-}
foo = id @(NatIndexed 4)
```

This results in the following LH error:

```
    • Specified type does not refine Haskell type for `HelloGADT.foo` (Plugged Init types old)
The Liquid type
.
    (HelloGADT.NatIndexed {4}) -> (HelloGADT.NatIndexed {4})
.
is inconsistent with the Haskell type
.
    HelloGADT.NatIndexed 4 -> HelloGADT.NatIndexed 4
.
defined at src/HelloGADT.hs:12:1-3
.
Specifically, the Liquid component
.
    {4}
.
is inconsistent with the Haskell component
.
    GHC.Types.Int
.

HINT: Use the hole '_' instead of the mismatched component (in the Liquid specification)
    • 
   |
11 | {-@ foo :: NI4 -> NI4 @-}
   |            ^^^^^^^^^^^
```

It seems that `NatIndexed 4` is parsed not as a Haskell type, but as some kind of LH expression.

**Comment by ranjitjhala** [comment]
Yes that’s what is happening; however this code should work treating NI4 as
a plain Haskell type alias?

**Comment by gergoerdi** [comment]
Wait are you asking me? I'm a bit confused about where this ticket is standing...

**Comment by ranjitjhala** [comment]
My apologies - I meant

1. this particular alias form is not currently supported by LH but
2. The code should work if the alias was a plain GHC alias (so there is no
need for LH to specifically support this use case.)

My question was: do you agree with 1 and 2? Apologies for being unclear!
Ranjit

**Comment by gergoerdi** [comment]
Ah, so the recommended workaround is to simply write:

```
{-# LANGUAGE DataKinds #-}
import Prelude
import GHC.TypeLits

data NatIndexed (n :: Nat)

type NI4 = NatIndexed 4 -- A Haskell, not LH, type synonym

{-@ foo :: NI4 -> NI4 @-}
foo = id @(NatIndexed 4)
```

I can confirm that this workaround works -- it's quite embarrassing that I didn't think of it :D

**Comment by gergoerdi** [comment]
Unfortunately, it seems the workaround doesn't quite cover all areas where LH has trouble with these type-level numerals.

```
type N0 = 0 :: Nat
{-@
data Vec n a where
    VNil :: Vec N0 a
@-}
data Vec n a where
    VNil :: Vec 0 a
```

This fails with:

```
    • Specified type does not refine Haskell type for `HelloVec.VNil` (Plugged Init types new)
The Liquid type
.
    k -> n -> a -> (HelloVec.Vec GHC.Types.Int a)
.
is inconsistent with the Haskell type
.
    k -> n -> a -> HelloVec.Vec 0 a
.
defined at /home/cactus/prog/haskell/liquid-haskell/clash/calculator/src/HelloVec.hs:35:5-20
.
Specifically, the Liquid component
.
    GHC.Types.Int
.
is inconsistent with the Haskell component
.
    k
.
```

and the suggestion doesn't work either, since this:

```
{-@
data Vec n a where
    VNil :: Vec _ a
@-}

data Vec n a where
    VNil :: Vec 0 a
```

fails with a similar error:

```
The Liquid type
.
    k -> n -> a -> (HelloVec.Vec _ a)
.
is inconsistent with the Haskell type
.
    k -> n -> a -> HelloVec.Vec 0 a
.
defined at /home/cactus/prog/haskell/liquid-haskell/clash/calculator/src/HelloVec.hs:35:5-20
.
Specifically, the Liquid component
.
    a
.
is inconsistent with the Haskell component
.
    GHC.Types.Int
.
```

**Comment by nikivazou** [comment]
@gergoerdi is this still a problem? 
The following is SAFE for me, on develop branch. 

```
{-@
data Vec n a where
    VNil :: Vec _ a
@-}

data Vec n a where
    VNil :: Vec 0 a
```

**Comment by kleinreact** [comment]
The following seems to work and helped me in getting some understanding of the underlying problem:
```haskell
type Zero = 0 :: Nat

{-@
data Vec where
  Nil :: Vec Zero a
  Cons :: a -> Vec n a -> Vec m a
@-}

data Vec n a where
  Nil :: Vec Zero a
  Cons :: n ~ m + 1 => a -> Vec m a -> Vec n a

{-@ measure size @-}
{-@ size :: Vec n a -> Nat @-}
size :: Vec n a -> Int
size Nil        = 0
size (Cons _ x) = 1 + size x
```

As far as I understand (and please correct me, if I got something wrong), LH has no knowledge about type level naturals and thus has a more limited type view than GHC. E.g., if we consider the usual `Cons` constructor definition for indexed vectors
```haskell
Cons :: a -> Vec n a -> Vec (n + 1) a
```
then LH threats `n` as a type variable, but also `n + 1` just as another one. Hence, LH is not aware of the semantic relation between the two.

The trick that I'm using above is making LH's ignorance of that type relation more explicit on the Haskell side as well, via shifting the relation into an explicit constraint in the constructor upfront. LH won't see that constraint either, hence it's type view and the Haskell definition now coincide. The other modification required here, is that we also need to drop the type parameters from the `data Vec` declaration in LH, as it seems like LH can't relate that type arguments with their usage in the constructors either, causing a LH type error, if they are kept. That's my intuition, at least.

So far so good. The example above now allows us to utilize a shared definition for indexed vectors without the need of restricting the usage of type level naturals on the Haskell side. With that at hand, we  can at least use the usual `size` measure definition, to make LH aware of the inductive structure of `Vec`. Hence, we basically re-establish the semantics of the type level naturals in LH this way, which is kind of to be expected, because the way type level naturals work in Haskell won't give any straightforward insights on the inductive structure of `Vec` anyway.

Nevertheless, also adding native support for type level naturals in LH would be a valuable extension, but that's kind of a different challenge, I guess.
